### ISSUE TITLE
Feature Request: Add the "stream" parameter which can be found inside normal requests.

### ISSUE THREAD

**Issue by VCDEV1** [yifeikong/curl_cffi#26]
First of all thank you for fixing #22 I think it is not available in pypi yet. I even tried to install it manually by doing `pip install git+https://github.com/yifeikong/curl_cffi` but it seems to give the below error on a linux (ubuntu) platform:
```
$ pip3 install git+https://github.com/yifeikong/curl_cffi
Collecting git+https://github.com/yifeikong/curl_cffi
  Cloning https://github.com/yifeikong/curl_cffi to /tmp/pip-req-build-gisjnq_5
  Running command git clone -q https://github.com/yifeikong/curl_cffi /tmp/pip-req-build-gisjnq_5
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Requirement already satisfied: cffi>=1.0.0 in ./.local/lib/python3.8/site-packages (from curl-cffi==0.3.8) (1.15.1)
Requirement already satisfied: pycparser in ./.local/lib/python3.8/site-packages (from cffi>=1.0.0->curl-cffi==0.3.8) (2.21)
Building wheels for collected packages: curl-cffi
  Building wheel for curl-cffi (PEP 517) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3 /tmp/tmpmigkk7kk build_wheel /tmp/tmp_eqeh8pr
       cwd: /tmp/pip-req-build-gisjnq_5
  Complete output (110 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-cpython-38
  creating build/lib.linux-x86_64-cpython-38/curl_cffi
  copying curl_cffi/build.py -> build/lib.linux-x86_64-cpython-38/curl_cffi
  copying curl_cffi/__init__.py -> build/lib.linux-x86_64-cpython-38/curl_cffi
  copying curl_cffi/_const.py -> build/lib.linux-x86_64-cpython-38/curl_cffi
  creating build/lib.linux-x86_64-cpython-38/curl_cffi/requests
  copying curl_cffi/requests/headers.py -> build/lib.linux-x86_64-cpython-38/curl_cffi/requests
  copying curl_cffi/requests/cookies.py -> build/lib.linux-x86_64-cpython-38/curl_cffi/requests
  copying curl_cffi/requests/session.py -> build/lib.linux-x86_64-cpython-38/curl_cffi/requests
  copying curl_cffi/requests/__init__.py -> build/lib.linux-x86_64-cpython-38/curl_cffi/requests
  copying curl_cffi/requests/errors.py -> build/lib.linux-x86_64-cpython-38/curl_cffi/requests
  running egg_info
  writing curl_cffi.egg-info/PKG-INFO
  writing dependency_links to curl_cffi.egg-info/dependency_links.txt
  writing requirements to curl_cffi.egg-info/requires.txt
  writing top-level names to curl_cffi.egg-info/top_level.txt
  reading manifest file 'curl_cffi.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  /tmp/pip-build-env-bx5qgl50/overlay/lib/python3.8/site-packages/setuptools/config/pyprojecttoml.py:108: _BetaConfiguration: Support for `[tool.setuptools]` in `pyproject.toml` is still *beta*.
    warnings.warn(msg, _BetaConfiguration)
  warning: no files found matching 'curl_cffi/cacert.pem'
  warning: no files found matching 'curl_cffi/*.dll'
  adding license file 'LICENSE'
  writing manifest file 'curl_cffi.egg-info/SOURCES.txt'
  copying curl_cffi/cdef.c -> build/lib.linux-x86_64-cpython-38/curl_cffi
  copying curl_cffi/shim.c -> build/lib.linux-x86_64-cpython-38/curl_cffi
  creating build/lib.linux-x86_64-cpython-38/curl_cffi/include
  copying curl_cffi/include/shim.h -> build/lib.linux-x86_64-cpython-38/curl_cffi/include
  creating build/lib.linux-x86_64-cpython-38/curl_cffi/include/curl
  copying curl_cffi/include/curl/Makefile.am -> build/lib.linux-x86_64-cpython-38/curl_cffi/include/curl
  copying curl_cffi/include/curl/curl.h -> build/lib.linux-x86_64-cpython-38/curl_cffi/include/curl
  copying curl_cffi/include/curl/curlver.h -> build/lib.linux-x86_64-cpython-38/curl_cffi/include/curl
  copying curl_cffi/include/curl/easy.h -> build/lib.linux-x86_64-cpython-38/curl_cffi/include/curl
  copying curl_cffi/include/curl/header.h -> build/lib.linux-x86_64-cpython-38/curl_cffi/include/curl
  copying curl_cffi/include/curl/mprintf.h -> build/lib.linux-x86_64-cpython-38/curl_cffi/include/curl
  copying curl_cffi/include/curl/multi.h -> build/lib.linux-x86_64-cpython-38/curl_cffi/include/curl
  copying curl_cffi/include/curl/options.h -> build/lib.linux-x86_64-cpython-38/curl_cffi/include/curl
  copying curl_cffi/include/curl/stdcheaders.h -> build/lib.linux-x86_64-cpython-38/curl_cffi/include/curl
  copying curl_cffi/include/curl/system.h -> build/lib.linux-x86_64-cpython-38/curl_cffi/include/curl
  copying curl_cffi/include/curl/typecheck-gcc.h -> build/lib.linux-x86_64-cpython-38/curl_cffi/include/curl
  copying curl_cffi/include/curl/urlapi.h -> build/lib.linux-x86_64-cpython-38/curl_cffi/include/curl
  running build_ext
  generating cffi module 'build/temp.linux-x86_64-cpython-38/curl_cffi._wrapper.c'
  creating build/temp.linux-x86_64-cpython-38
  building 'curl_cffi._wrapper' extension
  creating build/temp.linux-x86_64-cpython-38/build
  creating build/temp.linux-x86_64-cpython-38/build/temp.linux-x86_64-cpython-38
  creating build/temp.linux-x86_64-cpython-38/curl_cffi
  x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -Icurl_cffi/include -I/usr/include/python3.8 -c build/temp.linux-x86_64-cpython-38/curl_cffi._wrapper.c -o build/temp.linux-x86_64-cpython-38/build/temp.linux-x86_64-cpython-38/curl_cffi._wrapper.o
  x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -Icurl_cffi/include -I/usr/include/python3.8 -c curl_cffi/shim.c -o build/temp.linux-x86_64-cpython-38/curl_cffi/shim.o
  x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 build/temp.linux-x86_64-cpython-38/build/temp.linux-x86_64-cpython-38/curl_cffi._wrapper.o build/temp.linux-x86_64-cpython-38/curl_cffi/shim.o -L/usr/local/lib -L/usr/lib -lcurl-impersonate-chrome -o build/lib.linux-x86_64-cpython-38/curl_cffi/_wrapper.abi3.so
  /usr/bin/ld: cannot find -lcurl-impersonate-chrome
  collect2: error: ld returned 1 exit status
  /tmp/pip-build-env-bx5qgl50/overlay/lib/python3.8/site-packages/setuptools/command/build_py.py:202: SetuptoolsDeprecationWarning:     Installing 'curl_cffi.include' as data is deprecated, please list it in `packages`.
      !!


      ############################
      # Package would be ignored #
      ############################
      Python recognizes 'curl_cffi.include' as an importable package,
      but it is not listed in the `packages` configuration of setuptools.

      'curl_cffi.include' has been automatically added to the distribution only
      because it may contain data files, but this behavior is likely to change
      in future versions of setuptools (and therefore is considered deprecated).

      Please make sure that 'curl_cffi.include' is included as a package by using
      the `packages` configuration field or the proper discovery methods
      (for example by using `find_namespace_packages(...)`/`find_namespace:`
      instead of `find_packages(...)`/`find:`).

      You can read more about "package discovery" and "data files" on setuptools
      documentation page.


  !!

    check.warn(importable)
  /tmp/pip-build-env-bx5qgl50/overlay/lib/python3.8/site-packages/setuptools/command/build_py.py:202: SetuptoolsDeprecationWarning:     Installing 'curl_cffi.include.curl' as data is deprecated, please list it in `packages`.
      !!


      ############################
      # Package would be ignored #
      ############################
      Python recognizes 'curl_cffi.include.curl' as an importable package,
      but it is not listed in the `packages` configuration of setuptools.

      'curl_cffi.include.curl' has been automatically added to the distribution only
      because it may contain data files, but this behavior is likely to change
      in future versions of setuptools (and therefore is considered deprecated).

      Please make sure that 'curl_cffi.include.curl' is included as a package by using
      the `packages` configuration field or the proper discovery methods
      (for example by using `find_namespace_packages(...)`/`find_namespace:`
      instead of `find_packages(...)`/`find:`).

      You can read more about "package discovery" and "data files" on setuptools
      documentation page.


  !!

    check.warn(importable)
  error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
  ----------------------------------------
  ERROR: Failed building wheel for curl-cffi
Failed to build curl-cffi
ERROR: Could not build wheels for curl-cffi which use PEP 517 and cannot be installed directly
```
However I hope it will be fixed soon when the next version is pushed to pypi!

**This issue is about the `stream=True/False` parameter which can be found inside normal requests package. If it is possible can you add it to this please so I can perform operations like below:**
```
import requests

url="https://videos-fastly-studiodrm.jwpsrv.com/6411bcd6_0x344cee0abd6d9a757e2ea355371aba081c4aa3d3/site/Y21kWwti/media/UFCQHrhu/version/dozELTrB/manifest-34126610,34126611,34126612,34126615,34126616,34126617.ism/dash/manifest-34126610,34126611,34126612,34126615,34126616,34126617-audio_eng=112019.dash"
headers = {
    'authority': 'videos-fastly-studiodrm.jwpsrv.com',
    'accept': '*/*',
    'accept-language': 'en-GB,en-US;q=0.9,en;q=0.8',
    'cache-control': 'no-cache',
    'origin': 'https://juanflix.com.ph',
    'pragma': 'no-cache',
    'referer': 'https://juanflix.com.ph/',
    'sec-ch-ua': '"Google Chrome";v="111", "Not(A:Brand";v="8", "Chromium";v="111"',
    'sec-ch-ua-mobile': '?0',
    'sec-ch-ua-platform': '"Windows"',
    'sec-fetch-dest': 'empty',
    'sec-fetch-mode': 'cors',
    'sec-fetch-site': 'cross-site',
    'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36',
}
proxies={}
raw = requests.get(url=url, headers=headers, proxies=proxies, stream=True, verify=False)
if raw.status_code == 200 or raw.status_code == 206:
	for chunk in raw.iter_content(chunk_size=4096):
		if chunk:
			print("chunk found")
			# Do something
```

Thanks again and I am really sorry for disturbing you again and again.

**Comment by yifeikong** [comment]
First of all, please refer to the original issue for followup questions, it's more clear to discuss one topic in the same issue. I'll still be notified even if it's closed :P

Second, pypi package has been updated. Reinstall with `--upgrade`.

Last but not least. `stream` is a cool and must-have feature. I'll need some time to investigate `requests`'s interfaces and `libcurl`'s internal mechanism. Stay tuned.

**Comment by VCDEV1** [comment]
Thank you so much!

**Comment by qqqitianyu** [comment]
> stream=True/False

您好，我也希望您能尽快开发stream这个参数

**Comment by yifeikong** [comment]
最近又研究了下 libcurl 的 API，发现在**同步代码**里加上 stream 这个参数不太现实，因为 libcurl 只支持回调，没有遍历的 API。不过，有两个 workaround：

1. 可以使用 content_callback，传一个回调函数进去；
2. asyncio，理论上来说是可以支持 stream 的，不过我还没时间弄。

**Comment by VCDEV1** [comment]
@yifeikong Since asyncio is supported in the library now, will this be possible to add through asyncio?

**Comment by yifeikong** [comment]
Yes, it's possible to support stream via `asyncio.Queue`. It should be added in the next minor version.

**Comment by 1024bit** [comment]
stream这个功能很重要，希望作者早日整合到源码中。

**Comment by yifeikong** [comment]
Still working on this, in the meanwhile, you can use the `content_callback` parameter for callback style streaming.

```py
from curl_cffi import requests

def callback(data: bytes):
    print(data)

requests.get(url, content_callback=callback)
```

**Comment by yifeikong** [comment]
Added `stream=True` support in `v0.5.10b1`, tests yet to be added, but it seems to be working.

```py
with requests.Session() as s:
    print("\n======================================================================")
    print("Iterating over chunks")
    print("=====================================================================\n")
    r = s.get(URL, stream=True)
    for chunk in r.iter_content():
        print("CHUNK", chunk)
    r.close()

async def async_examples():
    async with requests.AsyncSession() as s:
        print("\n====================================================================")
        print("Using asyncio")
        print("====================================================================\n")
        r = await s.get(URL, stream=True)
        async for chunk in r.aiter_content():
            print("CHUNK", chunk)
        await r.aclose()
```

More examples can be found [here](https://github.com/yifeikong/curl_cffi/blob/master/examples/stream.py).

**Comment by hlohaus** [comment]
@yifeikong

Response status is not loaded:

curl_cffi.requests.errors.RequestsError: HTTP Error 0: OK

Example:
```
          response = await session.post(f"{url}/api/chat-process", json=data, headers=headers, stream=True)
          response.raise_for_status()
```

**Comment by yifeikong** [comment]
@hlohaus Please see if the wheels [here](https://github.com/yifeikong/curl_cffi/actions/runs/6371769166) in the artifacts section are working.

**Comment by hlohaus** [comment]
works so far:
```
        async with AsyncSession(impersonate="chrome110", verify=False) as session:
            async with await session.post(f"{url}/api/chat-process", stream=True) as response:
                response.raise_for_status()
                print("Empty:", response.text)
                async for line in response.aiter_lines():
                    print(line)
```
But response.text is empty.

Can you implement this interface:
```
        # new StreamSession with auto stream=True
        async with StreamSession(impersonate="chrome110", verify=False) as session:
             # session.post:
             # - has no await
             # - is not async and returns a "RequestSession" with __aenter__ and __enter__
             # - " __enter__" returns a "RequestSession" without async methods
             # - no stream=True
            async with session.post(f"{url}/api/chat-process") as response:
                response.raise_for_status()
                content_type = response.headers.get('content-type')
                if content_type == "application/octet-stream":
                    async for line in response.iter_lines():
                        print("Line:", line)
                else:
                    # text() and json() method with await:
                    text = await response.text()
                    print("Response:", text)
```

**Comment by hlohaus** [comment]
I found a other bug. If the response has no content:
```
async def run():
    async with AsyncSession(impersonate="chrome107") as session:
        response = await session.get("https://httpbin.org/status/202", stream=True)
        print(response.status_code)
asyncio.run(run())
```
We have a endless loop.

**Comment by yifeikong** [comment]
> works so far:
> 
> ```python
>         async with AsyncSession(impersonate="chrome110", verify=False) as session:
>             async with await session.post(f"{url}/api/chat-process", stream=True) as response:
>                 response.raise_for_status()
>                 print("Empty:", response.text)
>                 async for line in response.aiter_lines():
>                     print(line)
> ```
> 
> But response.text is empty.
> 
> Can you implement this interface:
> 
> ```python
>         # new StreamSession with auto stream=True
>         async with StreamSession(impersonate="chrome110", verify=False) as session:
>              # session.post:
>              # - has no await
>              # - is not async and returns a "RequestSession" with __aenter__ and __enter__
>              # - " __enter__" returns a "RequestSession" without async methods
>              # - no stream=True
>             async with session.post(f"{url}/api/chat-process") as response:
>                 response.raise_for_status()
>                 content_type = response.headers.get('content-type')
>                 if content_type == "application/octet-stream":
>                     async for line in response.iter_lines():
>                         print("Line:", line)
>                 else:
>                     # text() and json() method with await:
>                     text = await response.text()
>                     print("Response:", text)
> ```

I think what you described is the [aiohttp API](https://docs.aiohttp.org/en/stable/index.html), apart from the "no `stream=True`" part, it should be able to be implemented with a simple wrapper. 

The underlying `async with`-less API will still be kept, since

1. I'm not a huge fan of nested `with` blocks
2. As pointed out in the [`httpx` docs](https://www.python-httpx.org/async/#streaming-responses), `with` is not practical under some situation.

And thank you for reporting the bugs.

**Comment by yifeikong** [comment]
Here you go:

```py
async with s.stream("GET", URL) as r:
    print(await r.atext())

async with s.stream("GET", URL) as r:
    async for chunk in r.aiter_content():
        print("Status: ", r.status_code)
        assert r.status_code == 200
        print("CHUNK", chunk)
```

[Wheels](https://github.com/yifeikong/curl_cffi/actions/runs/6375883246).

**Comment by hlohaus** [comment]
```
    async with AsyncSession(impersonate="chrome107", headers={"Working": "Yes"}) as session:
        async with session.stream("GET", "https://httpbin.org/stream/20") as response:
            response.raise_for_status()
            async for line in response.aiter_lines():
                print(json.loads(line)["headers"]["Working"])
```
```
Yes
Yes
Yes
Yes
Yes
Yes
Yes
Yes
Yes
Yes
...
```

**Comment by yifeikong** [comment]
Available as `0.5.10b2`.

**Comment by yifeikong** [comment]
Landed in 0.5.10
